### PR TITLE
accelerate scripts for question answering with no trainer

### PR DESCRIPTION
**Pull request by theainerd**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors which may be interested in your PR.
@sgugger

### CODE REVIEW

File: examples/question-answering/run_qa_beam_search_no_trainer.py [hunk 1]
```diff
@@ -0,0 +1,808 @@
...
+
+        Args:
+            start_or_end_logits(:obj:`tensor`):
+                This is the output predictions of the model. We can only enter either start or end logits.
+            eval_dataset: Evaluation dataset
+            max_len(:obj:`int`):
+                The maximum length of the output tensor. ( See the model.eval() part for more details )
+        """
+
+        step_size = 0
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search_no_trainer.py:609]:
```suggestion
        step = 0
```
It's not really a size, more of an index. I think `step` is descriptive enough.

File: examples/question-answering/run_qa_beam_search_no_trainer.py [hunk 2]
```diff
@@ -0,0 +1,808 @@
...
+        logits_concat = np.full((len(dataset), max_len), -100, dtype=np.float32)
+        # Now since we have create an array now we will populate it with the outputs gathered using accelerator.gather
+        for i, output_logit in enumerate(start_or_end_logits):  # populate columns
+            # We have to fill it such that we have to take the whole tensor and replace it on the newly created array
+            # And after every iteration we have to change the step_size
+
+            batch_size = output_logit.shape[0]
+            cols = output_logit.shape[1]
+            logits_concat[step_size : step_size + batch_size, :cols] = output_logit
+            step_size = batch_size
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search_no_trainer.py:620]:
```suggestion
            logits_concat[step : step + batch_size, :cols] = output_logit
            step += batch_size
```
step needs to move by batch_size, not be equal to batch_size.

File: examples/question-answering/run_qa_beam_search_no_trainer.py [hunk 3]
```diff
@@ -0,0 +1,808 @@
...
+
+        max_len = max([x.shape[1] for x in all_end_top_log_probs])  # Get the max_length of the tensor
+
+        # concatenate all numpy arrays collected above
+        start_top_log_probs_concat = create_and_fill_np_array(all_start_top_log_probs, test_dataset, max_len)
+        start_top_index_concat = create_and_fill_np_array(all_start_top_index, test_dataset, max_len)
+        end_top_log_probs_concat = create_and_fill_np_array(all_end_top_log_probs, test_dataset, max_len)
+        end_top_index_concat = create_and_fill_np_array(all_end_top_index, test_dataset, max_len)
+
+        del start_top_log_probs
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search_no_trainer.py:784]: Maybe add a comment above on what this is doing?

File: examples/question-answering/run_qa_beam_search_no_trainer.py [hunk 4]
```diff
@@ -0,0 +1,808 @@
...
+
+        max_len = max([x.shape[1] for x in all_end_top_log_probs])  # Get the max_length of the tensor
+
+        # concatenate all numpy arrays collected above
+        start_top_log_probs_concat = create_and_fill_np_array(all_start_top_log_probs, eval_dataset, max_len)
+        start_top_index_concat = create_and_fill_np_array(all_start_top_index, eval_dataset, max_len)
+        end_top_log_probs_concat = create_and_fill_np_array(all_end_top_log_probs, eval_dataset, max_len)
+        end_top_index_concat = create_and_fill_np_array(all_end_top_index, eval_dataset, max_len)
+
+        del start_top_log_probs
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search_no_trainer.py:730]: Maybe add a comment on what this is doing?

File: examples/question-answering/run_qa_no_trainer.py [hunk 5]
```diff
@@ -0,0 +1,767 @@
...
+        logits_concat = np.full((len(dataset), max_len), -100, dtype=np.float64)
+        # Now since we have create an array now we will populate it with the outputs gathered using accelerator.gather
+        for i, output_logit in enumerate(start_or_end_logits):  # populate columns
+            # We have to fill it such that we have to take the whole tensor and replace it on the newly created array
+            # And after every iteration we have to change the step_size
+
+            batch_size = output_logit.shape[0]
+            cols = output_logit.shape[1]
+            logits_concat[step_size : step_size + batch_size, :cols] = output_logit
+            step_size = batch_size
```
Comment (sgugger) [inline examples/question-answering/run_qa_no_trainer.py:623]: Same comments as above here.

File: examples/question-answering/run_qa_no_trainer.py [hunk 6]
```diff
@@ -0,0 +1,767 @@
...
+        else:
+            formatted_predictions = [{"id": k, "prediction_text": v} for k, v in predictions.items()]
+
+        references = [{"id": ex["id"], "answers": ex[answer_column_name]} for ex in examples]
+        return EvalPrediction(predictions=formatted_predictions, label_ids=references)
+
+    metric = load_metric("squad_v2" if args.version_2_with_negative else "squad")
+
+    def compute_metrics(p: EvalPrediction):
+        return metric.compute(predictions=p.predictions, references=p.label_ids)
```
Comment (sgugger) [inline examples/question-answering/run_qa_no_trainer.py:597]: Same comments as above here.

File: examples/question-answering/run_qa_no_trainer.py [hunk 7]
```diff
@@ -0,0 +1,767 @@
...
+                outputs = model(**batch)
+                start_logits = outputs.start_logits
+                end_logits = outputs.end_logits
+
+                if not args.pad_to_max_length:  # necessary to pad predictions and labels for being gathered
+                    start_logits = accelerator.pad_across_processes(start_logits, dim=1, pad_index=-100)
+                    end_logits = accelerator.pad_across_processes(end_logits, dim=1, pad_index=-100)
+
+                all_start_logits.append(accelerator.gather(start_logits.cpu().numpy()))
+                all_end_logits.append(accelerator.gather(end_logits.cpu().numpy()))
```
Comment (sgugger) [inline examples/question-answering/run_qa_no_trainer.py:711]: Same comments as above here.

File: examples/question-answering/run_qa_no_trainer.py [hunk 8]
```diff
@@ -0,0 +1,767 @@
...
+                start_logits = outputs.start_logits
+                end_logits = outputs.end_logits
+
+                if not args.pad_to_max_length:  # necessary to pad predictions and labels for being gathered
+                    start_logits = accelerator.pad_across_processes(start_logits, dim=1, pad_index=-100)
+                    end_logits = accelerator.pad_across_processes(start_logits, dim=1, pad_index=-100)
+
+                all_start_logits.append(accelerator.gather(start_logits.cpu().numpy()))
+                all_end_logits.append(accelerator.gather(end_logits.cpu().numpy()))
+
```
Comment (sgugger) [inline examples/question-answering/run_qa_no_trainer.py:744]: Same comments as above here.

File: examples/question-answering/run_qa_beam_search_no_trainer.py [hunk 9]
```diff
@@ -0,0 +1,808 @@
...
+                    start_top_log_probs = accelerator.pad_across_processes(start_top_log_probs, dim=1, pad_index=-100)
+                    start_top_index = accelerator.pad_across_processes(start_top_index, dim=1, pad_index=-100)
+                    end_top_log_probs = accelerator.pad_across_processes(end_top_log_probs, dim=1, pad_index=-100)
+                    end_top_index = accelerator.pad_across_processes(end_top_index, dim=1, pad_index=-100)
+
+                all_start_top_log_probs.append(accelerator.gather(start_top_log_probs.cpu().numpy()))
+                all_start_top_index.append(accelerator.gather(start_top_index.cpu().numpy()))
+                all_end_top_log_probs.append(accelerator.gather(end_top_log_probs.cpu().numpy()))
+                all_end_top_index.append(accelerator.gather(end_top_index.cpu().numpy()))
+                all_cls_logits.append(accelerator.gather(cls_logits.cpu().numpy()))
```
Comment (sgugger) [inline examples/question-answering/run_qa_beam_search_no_trainer.py:774]: Same comments as above here.

**Review by sgugger — APPROVED** [review]
Thanks for your changes! I have a few more comments and this should be ready to merge!

File: examples/question-answering/run_qa_beam_search_no_trainer.py (hunk 4 above)
Comment (theainerd) [reply examples/question-answering/run_qa_beam_search_no_trainer.py:730]: Just deleting it to be on the safe side to save some memory but not sure how much this helps.

**Pull request closed without merging**